src/infra/core/src/repos/dataset_repository_s3.rs

```diff
@@ -26,76 +25,105 @@ use crate::utils::s3_context::S3Context;
 pub struct DatasetRepositoryS3 {
     s3_context: S3Context,
     current_account_subject: Arc<CurrentAccountSubject>,
+    multi_tenant: bool,
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
 impl DatasetRepositoryS3 {
-    pub fn new(s3_context: S3Context, current_account_subject: Arc<CurrentAccountSubject>) -> Self {
+    pub fn new(
+        s3_context: S3Context,
+        current_account_subject: Arc<CurrentAccountSubject>,
+        multi_tenant: bool,
+    ) -> Self {
         Self {
             s3_context,
             current_account_subject,
+            multi_tenant,
         }
     }
 
     async fn get_dataset_impl(
         &self,
-        dataset_alias: &DatasetAlias,
+        dataset_id: &DatasetID,
     ) -> Result<impl Dataset, InternalError> {
-        assert!(
-            !dataset_alias.is_multi_tenant(),
-            "Multi-tenancy is not yet supported by S3 repo"
-        );
-
         let s3_context = self
             .s3_context
-            .sub_context(&format!("{}/", &dataset_alias.dataset_name));
+            .sub_context(&format!("{}/", &dataset_id.cid.to_string()));
 
         DatasetFactoryImpl::get_s3_from_context(s3_context).await
     }
 
-    async fn delete_dataset_s3_objects(
-        &self,
-        dataset_alias: &DatasetAlias,
-    ) -> Result<(), InternalError> {
-        assert!(
-            !dataset_alias.is_multi_tenant(),
-            "Multi-tenancy is not yet supported by S3 repo"
-        );
-        let dataset_key_prefix = self.s3_context.get_key(&dataset_alias.dataset_name);
+    async fn delete_dataset_s3_objects(&self, dataset_id: &DatasetID) -> Result<(), InternalError> {
+        let dataset_key_prefix = self.s3_context.get_key(&dataset_id.cid.to_string());
         self.s3_context.recursive_delete(dataset_key_prefix).await
     }
 
-    async fn move_bucket_items_on_dataset_rename(
+    async fn resolve_dataset_alias(
         &self,
-        old_alias: &DatasetAlias,
-        new_alias: &DatasetAlias,
-    ) -> Result<(), MoveBucketItemsOnRenameError> {
-        assert!(
-            !new_alias.is_multi_tenant(),
-            "Multi-tenancy is not yet supported by S3 repo"
-        );
-
-        let new_key_prefix = self.s3_context.get_key(&new_alias.dataset_name);
-        if self
-            .s3_context
-            .bucket_path_exists(new_key_prefix.as_str())
-            .await?
-        {
-            return Err(MoveBucketItemsOnRenameError::NameCollision(
-                NameCollisionError {
-                    alias: new_alias.clone(),
-                },
-            ));
+        dataset: &dyn Dataset,
+    ) -> Result<DatasetAlias, InternalError> {
+        match dataset.as_info_repo().get("alias").await {
```

Comment:
why use `match` instead of `dataset.as_info_repo().get("alias").await.int_err()?` just like below?

Reply:
Simplified. This was copied & cut from a more complex match in Local FS